### PR TITLE
disable innerOuter schedule if iteration domain in inner reduction tv is not mapped to reduction domain in outer reduction tv

### PR DESCRIPTION
**Pull request by liqiangxl**
In #1023, the iteration domain in inner reduction tv is not mapped to reduction domain in outer reduction tv.
I can reproduce the error with a simple fusion:
```
Inputs:
  T0_g[ iS0{i0}, iS1{i2}, iS2{i3} ], float
Outputs:
  T3_g[ iS9{i0}, iS10{i2}, iS11{i3} ], float
  T4_g[ rS12{i0}, iS13{i2}, iS14{i3} ], float

%kernel_math {
T1_l[ iS3{i0}, iS4{i2}, rS5{i3} ]
   = reduction( T0_g[ iS0{i0}, iS1{i2}, iS2{i3} ], op = add, initial value = float(0), allreduce = false )
T2_l[ iS6{i0}, iS7{i2}, bS8{1} ]
   = broadcast( T1_l[ iS3{i0}, iS4{i2}, rS5{i3} ] )
T3_g[ iS9{i0}, iS10{i2}, iS11{i3} ]
   = T2_l[ iS6{i0}, iS7{i2}, bS8{1} ]
   + T0_g[ iS0{i0}, iS1{i2}, iS2{i3} ];
T4_g[ rS12{i0}, iS13{i2}, iS14{i3} ]
   = reduction( T0_g[ iS0{i0}, iS1{i2}, iS2{i3} ], op = add, initial value = float(0), allreduce = false )
}
}
```
In this fusion, inner reduction is `T1_g[ iS3{i0}, iS4{i2}, rS5{i3} ]`, outer reduction is `T4_g[ rS12{i0}, iS13{i2}, iS14{i3} ]`, The axis `iS4{i2}` in `T1_g` should be a reduction in `T4_g`, otherwise we can't do a partial outer reduction while iterating the iteration domain of inner reduction.

### CODE REVIEW

**Comment by liqiangxl** [comment]
!build

**Comment by liqiangxl** [comment]
!build

**Comment by liqiangxl** [comment]
!build

**Comment by xwang233** [comment]
ICYMI:

_@github-actions  nvfuser-ci/jit_codegen_diff_4/9 — Successful (w/ msg) in 56 minutes [Details](https://github.com/NVIDIA/Fuser/actions/runs/6424217913)_

See rendered HTML page at 
http://nv/e5M/nvfuser_github_ci/codegen_diff_p10125056_j70931434_1696538138698772815_codediff_3128102_48bda6c_custom_command_20231005_123930.html

**Comment by naoyam** [comment]
Do you think this is a fundamental constraint or we can extend the scheduler to support patterns like this? Disabling it for now is completely fine, just thinking if there's a way to support like this

File: csrc/scheduler/normalization_utils.cpp [hunk 1]
```diff
@@ -601,6 +601,41 @@ bool isConnectedOnlyThroughReductionProducer(
...
+    if (reference_tv->axis((int)i)->isReduction()) {
+      is_reduction[i] = true;
+    }
+  }
+  // check other inner reduction tvs, the corresponding axis should be
+  // reduction.
+  for (auto i : c10::irange(1, inner_reduction_tvs.size())) {
+    auto tv = inner_reduction_tvs[i];
+    for (const auto i : c10::irange(tv->nDims())) {
+      if (tv->axis((int)i)->isReduction() != is_reduction[i]) {
```
Comment (naoyam) [inline csrc/scheduler/normalization_utils.cpp:622]: I know at this point we already have checked all tvs have the same number of dimensions, but let's protect us from future stupid mistakes

```suggestion
      if (tv->axis((int)i)->isReduction() != is_reduction.at(i)) {
```

File: csrc/scheduler/normalization_utils.cpp [hunk 2]
```diff
@@ -601,6 +601,41 @@ bool isConnectedOnlyThroughReductionProducer(
...
+        return false;
+      }
+    }
+  }
+  // check outer reduction tvs, the corresponding axis should be iteration.
+  for (auto tv : outer_reduction_tvs) {
+    for (const auto i : c10::irange(tv->nDims())) {
+      if (tv->axis((int)i)->isIteration() != is_reduction[i]) {
```
Comment (naoyam) [inline csrc/scheduler/normalization_utils.cpp:630]: What happens if one domain is a broadcast in both inner and outer tensors. Do we support tensors with broadcast domains? At this this condition should just pass. Not sure if that's a problem.

Comment (liqiangxl) [reply csrc/scheduler/normalization_utils.cpp:630]: `checkIfReductionsAreInnerOuter` already ensures all the tensor domains are either iteration or reduction. Will reject if there is a broadcast domain.

**Comment by liqiangxl** [comment]
> Do you think this is a fundamental constraint or we can extend the scheduler to support patterns like this? Disabling it for now is completely fine, just thinking if there's a way to support like this

I believe we can accommodate this pattern, but it would require significant modifications to the existing scheduler. In the current code, we combine all iteration domains, start the schdule from a tensor domain of [I,R]. To facilitate this pattern, we would need to start from a domain of [I, I, R]. I'll create an issue tracking it.

File: csrc/scheduler/normalization_utils.cpp (hunk 2 above)
Comment (naoyam) [reply csrc/scheduler/normalization_utils.cpp:630]: OK, please assert that here.

**Comment by naoyam** [comment]
> > Do you think this is a fundamental constraint or we can extend the scheduler to support patterns like this? Disabling it for now is completely fine, just thinking if there's a way to support like this
> 
> I believe we can accommodate this pattern, but it would require significant modifications to the existing scheduler. In the current code, we combine all iteration domains, start the schdule from a tensor domain of [I,R]. To facilitate this pattern, we would need to start from a domain of [I, I, R]. I'll create an issue tracking it.

Makes sense.

**Comment by naoyam** [comment]
Why this test failed? 

https://github.com/NVIDIA/Fuser/actions/runs/6424129989

CC: @xwang233

**Comment by xwang233** [comment]
The error happened at run time, where the error message said the cuda driver is too old.

I think the root cause is we're building that example in a wrong way that cuda compatibility isn't handled correctly, which is a real issue for users.

Do you want to pause that job if that made you confused?

**Comment by naoyam** [comment]
> The error happened at run time, where the error message said the cuda driver is too old.
> 
> I think the root cause is we're building that example in a wrong way that cuda compatibility isn't handled correctly, which is a real issue for users.
> 
> Do you want to pause that job if that made you confused?

I think it's best to disable the test and create an issue to keep track of the build issue.

File: csrc/scheduler/normalization_utils.cpp [hunk 3]
```diff
@@ -601,6 +601,47 @@ bool isConnectedOnlyThroughReductionProducer(
...
+    const std::vector<TensorView*>& inner_reduction_tvs,
+    const std::vector<TensorView*>& outer_reduction_tvs) {
+  // set up reference, checkIfReductionsAreInnerOuter already ensures all the
+  // tensor domains are either iteration or reduction, so we can just use a
+  // vector of bool.
+  auto reference_tv = inner_reduction_tvs[0];
+  std::vector<bool> is_reduction(reference_tv->nDims(), false);
+  for (const auto i : c10::irange(reference_tv->nDims())) {
+    auto id = reference_tv->axis((int)i);
+    NVF_CHECK(!id->isBroadcast(), "Shouldn't have a broadcast domain.");
```
Comment (naoyam) [inline csrc/scheduler/normalization_utils.cpp:614]:
```suggestion
    NVF_CHECK(id->getIterationType() == IterationType::Iteration || id->getIterationType() == IterationType==Reduction, "Invalid iteration type: ", id->getIterationType());
```

**Review by naoyam — APPROVED** [review]
LGTM. Thanks for the fix

**Pull request merged**